### ISSUE TITLE
Implement and Test Different Gradient Descent in BOLT

### ISSUE THREAD

**Issue by Third-AI** [ThirdAILabs/Universe#105]
Implement and Test following Functionalities. Essentially add simple functionalities to 
void SparseLayer::UpdateParameters(float lr, uint32_t iter, float B1, float B2, float eps) 

- [ ]   Disable Adam and go to Plain SGD after a number of epoch or if the training loss does not jump by a relative amount

- [ ]   Custom optimizer, Keep two auxiliary variable: learning rate and previous gradient for every parameter.  if sign of current gradient  = sign of previous gradient -- learning rate = lambda x  learningrate (lambda > 1) else   learning rate = lambda^-1 x  learningrate 
- [ ] Some way to plug and play custom optimization rules

Remember, with large batch size gradient is way more reliable, so we can simply do plain SGD and it should be good.
